pallets/referrals/src/lib.rs

```diff
@@ -0,0 +1,573 @@
+
+		/// Maximum referral code length.
+		#[pallet::constant]
+		type CodeLength: Get<u32>;
+
+		/// Volume needed to next tier. If None returned, it is the last tier.
+		type TierVolume: GetByKey<Level, Option<Balance>>;
+
+		/// Seed amount that was sent to the reward pot.
+		#[pallet::constant]
+		type SeedNativeAmount: Get<u128>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		#[cfg(feature = "runtime-benchmarks")]
+		type BenchmarkHelper: BenchmarkHelper<Self::AssetId, Balance>;
+	}
+
+	/// Referral codes
+	/// Maps an account to a referral code.
+	#[pallet::storage]
+	#[pallet::getter(fn referral_account)]
+	pub(super) type ReferralCodes<T: Config> =
+		StorageMap<_, Blake2_128Concat, ReferralCode<T::CodeLength>, T::AccountId>;
+
+	/// Linked accounts.
+	/// Maps an account to a referral account.
+	#[pallet::storage]
+	#[pallet::getter(fn linked_referral_account)]
+	pub(super) type LinkedAccounts<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, T::AccountId>;
+
+	/// Shares per account.
+	#[pallet::storage]
+	#[pallet::getter(fn account_shares)]
+	pub(super) type Shares<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, Balance, ValueQuery>;
+
+	/// Total share issuance.
+	#[pallet::storage]
+	#[pallet::getter(fn total_shares)]
+	pub(super) type TotalShares<T: Config> = StorageValue<_, Balance, ValueQuery>;
+
+	/// Referer level and total accumulated rewards over time.
+	/// Maps referrer account to (Level, Balance). Level indicates current reward tier and Balance is used to unlock next tier level.
+	/// Dev note: we use OptionQuery here because this helps to easily determine that an account if referrer account.
+	#[pallet::storage]
+	#[pallet::getter(fn referrer_level)]
+	pub(super) type Referrer<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, (Level, Balance), OptionQuery>;
+
+	/// Asset tier information.
+	/// Maps (asset_id, level) to Tier which provides information about reward percentages.
+	#[pallet::storage]
+	#[pallet::getter(fn asset_tier)]
+	pub(super) type AssetTier<T: Config> =
+		StorageDoubleMap<_, Blake2_128Concat, T::AssetId, Blake2_128Concat, Level, Tier, OptionQuery>;
+
+	/// Information about assets that are currently in the rewards pot.
+	/// Used to easily determine list of assets that need to be converted.
+	#[pallet::storage]
+	#[pallet::getter(fn assets)]
+	pub(super) type Assets<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, ()>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Referral code has been registered.
+		CodeRegistered {
+			code: ReferralCode<T::CodeLength>,
+			account: T::AccountId,
+		},
+		/// Referral code has been linked to an account.
+		CodeLinked {
+			account: T::AccountId,
+			code: ReferralCode<T::CodeLength>,
+			referral_account: T::AccountId,
+		},
+		/// Asset has been converted to RewardAsset.
+		Converted {
+			from: T::AssetId,
```

Comment:
probably it would make sense to have "Amount" struct with asset id and value, so you would have only from & to fields here : Amount

Reply:
done